### ISSUE TITLE
Allow user to add headers to the ZAP API

### ISSUE THREAD

**Issue by psiinon** [zaproxy/zaproxy#4529]
As per https://groups.google.com/d/msg/zaproxy-users/UXzDPz9laA0/foAICdwFBgAJ
2 options come to mind:

1. Extending scripts to cover ZAP API responses
1. Adding the option for the user to explicitly set ZAP API headers

Its worth noting that changing the ZAP API security headers could make the API vulnerable to various attacks, which is why I'm a bit nervous about extending scripts to support them.

Thoughts?

**Comment by thc202** [comment]
Based on the use case I'd go with option 2.

**Comment by psiinon** [comment]
Yeah, then we can add explicit warnings - "Make sure you really know what you're doing" ;)

**Comment by cesarga** [comment]
For the CORS issue in particular, a checkbox toggle would be really helpful and more than enough.

You don't even have to use the "*" wildcard. If the request comes from one of the "Addresses permitted to use the API" it should set the _Access-Control-Allow-Origin_ header value to that origin.

For the moment I'm using these extensions for [Firefox](https://addons.mozilla.org/en-US/firefox/addon/cors-everywhere/) and [Chrome](https://chrome.google.com/webstore/detail/allow-control-allow-origi/nlfbmbojpeacfghkpbjhddihlkkiljbi?hl=en). And, if this feature isn't ready soon, I'll just proxy the ZAP API adding the header.

**Comment by Jiff21** [comment]
I'm trying to use https://github.com/zaproxy/zap-api-python to trigger tests via https://hub.docker.com/r/owasp/zap2docker-stable/ to integrate into AI but I can't find a way to add an Authorization header so it can access our dev environment instead of a login page. Just wanted to add a use case that fits this issues summary.

**Comment by thc202** [comment]
@Jiff21 that's a different use case, it's already possible to inject headers into the requests sent by ZAP, for example, using Replacer add-on [1] or HTTP Sender script. [2][3] (Both functionalities can be enabled through the ZAP API.)

[1] https://github.com/zaproxy/zap-extensions/wiki/HelpAddonsReplacerReplacer
[2] https://github.com/zaproxy/zap-core-help/wiki/HelpAddonsScriptsScripts
[3] https://github.com/zaproxy/community-scripts/tree/master/httpsender

**Comment by Jiff21** [comment]
Apologies if I'm missing something but how would you use any of those if you're just running a docker image of owasp/zap2docker-stable and triggering scans via baseline or zap cli commands. To add add-ons it seems like you'd modify the dockerfile, or is there documentation for how to add an add-on to the image?

My intention was to provide a different use case, since it's titled "Allow user to add headers to the ZAP API" it seemed like other use cases should be considered.

**Comment by thc202** [comment]
Both add-ons are included by default, so you can just use them (there are command line arguments [1] and ZAP API endpoints [2] to install add-ons though). For example, you can use the Phyton ZAP API client to set the replacer rule that injects the desired Authorization header:
```Python
zap.replacer.add_rule(description = "MyRuleAuthHeader", enabled = True, matchtype = "REQ_HEADER", matchregex = False, matchstring = "Authorization", replacement = "Auth Token")
```

This issue is about the headers returned by the ZAP API, your use case is to add headers to requests sent by ZAP (e.g. spider, active scanner) even if using the ZAP API.

[1] https://github.com/zaproxy/zap-core-help/wiki/HelpCmdline
[2] https://github.com/zaproxy/zaproxy/wiki/ApiGen_autoupdate

**Comment by VikashKrChoudhary** [comment]
The issue which I was facing was similar to the reported one where "Access-Control-Allow-Origin" set as "ZAP API" i.e. "Access-Control-Allow-Origin: ZAP API".
So as a workaround I excluded the endpoint from the proxy. Using option "Exclude From" -> Proxy.

I do understand that it can not be considered as an expected solution but yes this is how I use it as a workaround and it worked well.
Hope this helps!
